lib/core/src/Control/Monad/Util.hs

```diff
@@ -0,0 +1,37 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- |
+-- Copyright: © 2022 IOHK
+-- License: Apache-2.0
+--
+-- Monadic utility functions.
+--
+module Control.Monad.Util
+    ( applyNM
+    ) where
+
+import Prelude
+
+-- | Applies a monadic function @n@ times to an initial value.
+--
+-- Returns the final result, equivalent to the nth element of the following
+-- sequence:
+--
+-- @
+-- [ pure a
+-- , f a
+-- , (f <=< f) a
+-- , (f <=< f <=< f) a
+-- , (f <=< f <=< f <=< f) a
+-- , ...
+-- ]
+-- @
+--
+applyNM :: forall m a. Monad m => Int -> (a -> m a) -> a -> m a
+applyNM n next = loop n
+  where
+    loop :: Int -> a -> m a
+    loop !i !a
+        | i <= 0 = pure a
+        | otherwise = loop (i - 1) =<< next a
```

Comment:
```suggestion
applyNM :: forall m a. Monad m => Int -> (a -> m a) -> a -> m a
applyNM n next = (!! n) . iterate (next =<<)
```

[Wholemeal programming](https://stackoverflow.com/questions/6957270/what-is-wholemeal-in-functional-programming) 🤓

This version is "obviously correct" — I think QuickChecking this one is a bit overkill. If you really want to, checking the fast version against the "obviously correct" one may be worth doing, though.

Reply:
> This version is "obviously correct"

It's definitely more concise! (Though not sure about the performance.)

There is one difference though with this more concise implementation: it fails with a run-time error if `n` is less than `0`. (Whereas the original function treats negative values in the same way as functions like `take`.)

>  If you really want to, checking the fast version against the "obviously correct" one may be worth doing, though.

Thanks, that's a nice idea! I've added a commit that does this:  bda9108e4c6db0a4545ffdd6d7de79a836b9f8c0.